File: problems/leetcode/PathSum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public boolean hasPathSum(TreeNode root, int sum) {
        return root == null ? false : hasPathSum(root, sum, 0);
    }
    
    public boolean hasPathSum(TreeNode root, int sum, int curSum) {
        curSum += root.val;
        if (root.left == null && root.right == null) {
            return  (curSum == sum);
        }
        
        boolean left = root.left != null ? hasPathSum(root.left, sum, curSum) : false;
        boolean right = root.right != null ? hasPathSum(root.right, sum, curSum) : false;
        
        return left || right;
    }
}

File: problems/leetcode/path-sum.py
# https://leetcode.com/problems/path-sum/
class Solution:
    def hasPathSum(self, root: Optional[TreeNode], targetSum: int) -> bool:
        if not root:
            return False

        if not (root.left or root.right):
            return (targetSum - root.val) == 0

        return self.hasPathSum(root.left, targetSum - root.val) or self.hasPathSum(root.right, targetSum - root.val)
